### PR TITLE
Introduce Opt-in mode as an alternative configuration

### PR DESCRIPTION
**Pull request by theSteveMitchell**
Introduces a new configuration option, and a new mode of operating as a request filter. 

**Tl;DR:** 
If keycloak-api-rails is configured with `opt_in: true`, it will not validate all requests by default. Instead, it will allow controllers to opt-in to validation. 

**Use Case**
As our API expands, we have several clients/systems interacting with that API, and several different rules/modes of authentication. We see an increase in the number of controllers/routes that require a mode of authentication other than Keycloak (e.g. some use static API tokens, some use a different OAuth provider). We have a few issues that we've encountered:
 - The skip_paths list is growing, and it's challenging to maintain a list of many regexp. We find we have to write tests to validate the list of regexp meets our expectations
 - It's difficult to get granular control of validation. E.g we need to patch the middleware, which can be brittle.
 - Application errors are hidden by authentication errors. E.g. When we try to CURL a path that doesn't exist in our app, we get a 401 response instead of a 404

**Solution**
Keycloak-api-rails will continue to use the same middleware implementation unchanged, by default.

However, if it is configured with `opt-in: true`, that middleware will not be used, and instead Keycloak can be called explicitly. In other words, controllers/routes will **opt-in** to validation, rather than having to opt-out with a skip_path setting

Example Config:
```ruby
Keycloak.configure do |config|
  config.server_url = ENV["KEYCLOAK_SERVER_URL"]
  config.realm_id   = ENV["KEYCLOAK_REALM_ID"]
  config.logger     = Rails.logger
  config.opt_in     = true
end
```

When opt-in mode is used, Keycloak will not validate requests in the middleware layer. Application will need to authenticate with Keycloak by including the authentication module, and calling the helper (e.g. from a controller)

```ruby
class AuthenticatedController < ApplicationController
  include Keycloak::Authentication

  before_action :keycloak_authenticate, only: show

  def show
    # requires Keycloak authentication
    keycloak_id = Keycloak::Helper.current_user_id(request.env)
    User.active.find_by(keycloak_id: keycloak_id)
  end

  def index 
    # unauthenticated
  end
end
```

Nothing changes about the mechanism for validating the token itself.

### CODE REVIEW

**Review by looorent — APPROVED** [review]
Nice one, @theSteveMitchell !
I have faced the same issues. Thanks for your contribution!
It has been published as `0.12.0`.

**Comment by theSteveMitchell** [comment]
Awesome, thank you for the quick turnaround @looorent
